### PR TITLE
feat: self-identified account-link flow — token, lookup, request-email, validate (#2035)

### PR DESCRIPTION
**Pull request by TheTechArch**
Implements the authentication side of the self-identified (SI) account-link flow for #2035 — letting an authenticated person reclaim/link a legacy SI account by proving control of its registered email. Builds on #2025 (SI credential storage incl. email).

## Scope delivered

End-to-end on the authentication side:

1. **Link token (mint/validate)** — `SelfIdentifiedLinkTokenService` issues a short-lived, single-purpose JWT binding `from_party_uuid` (the SI user) to `to_party_uuid` (the authenticated person), with `purpose`/`jti` claims and a dedicated `iss`/`aud`. Signed via a dedicated certificate **seam** (`ISelfIdentifiedLinkTokenCertificateProvider`) kept separate from the OIDC signing keys, so the token can never be used for authentication. Interim impl borrows the existing provider; a dedicated cert is the documented follow-up.
2. **SI link-target lookup** — `IUserProfileService.GetSelfIdentifiedLinkTargetAsync` returns the SI user's `PartyUuid` + stored `Email`, or `null` for unknown/inactive/no-email (single null = no enumeration via branching).
3. **`POST .../selfidentified/link-request`** — looks up the SI user by username, mints the token, builds the access-management link, and sends the email **directly to the SI address** via a new thin Altinn Notifications client (`future/orders`, `RecipientEmail`, immediate delivery). Returns the **masked** recipient (`r*****@g****.com`) so the UI can show where it went; the full address never leaves authentication.
4. **`POST .../selfidentified/validate-link-token`** — validates the token, enforces **requester == consumer** (caller's party UUID must equal `to_party_uuid`), and returns `from_party_uuid` — the same `Guid` shape as `validate-credentials`, so access-management reuses `PostNewSelfIdentifiedUser(from, to)` unchanged.

Plus config/DI: `SelfIdentifiedLinkTokenSettings`, `SelfIdentifiedLinkSettings`, `PlatformSettings.ApiNotificationsEndpoint`, and registrations.

## Tests

Unit tests (no Docker): token mint/validate (issuer/audience/lifetime/signature/wrong-key), link orchestration (mint+send, masked-email return, no-op on unknown/failed send, mask edge cases), SI link-target lookup branches, and the `validate-link-token` controller behaviours (200/401/403/400).

## Deliberate decisions / notes

- **Masked-email return relaxes strict no-enumeration** — a UX requirement (show the recipient). Limited to authenticated callers; only a masked form is returned; rate-limiting is the real abuse control (tracked in #2035).
- **Idempotency** of the email send is a deterministic minute-bucketed key; verified against Notifications' `UNIQUE (idempotencyid, creatorname)` constraint, so it dedupes across pods.
- **Notification client registration is tolerant** of a missing endpoint so unrelated actions on the shared controller aren't affected; a send then fails gracefully.

## Out of scope / follow-ups (tracked in #2035)

- Dedicated link-token certificate (KeyVault) — wired through the existing seam.
- `jti` single-use store (token is replayable within its short TTL until then).
- Email localization (nb/nn/en).
- Access-management landing page + connection creation (separate repo).
- Infra: notifications scope on the platform access token + `ApiNotificationsEndpoint` per environment.

Refs #2035 · builds on #2025

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Self‑identified users can request account‑link emails and redeem short‑lived link tokens to create account connections; responses return masked recipient emails.
  * Platform notifications integration for sending link emails and direct creation of self‑identified connections in access management.
  * Configurable link/token settings and secure token minting/validation with certificate support.
  * Local credential lockout tracking to mitigate brute‑force attacks.

* **Tests**
  * Added unit tests for link request/redemption, token mint/validation, email masking, and lockout behaviors.

### CODE REVIEW

File: src/Authentication/Services/SelfIdentifiedLinkService.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                return null;
+            }
+
+            string token = await _linkTokenService.MintAsync(target.PartyUuid, toPartyUuid, cancellationToken);
+            string linkUrl = QueryHelpers.AddQueryString(_settings.AccessManagementLinkUrl, "token", token);
+            string body = BuildEmailBody(linkUrl);
+            string idempotencyId = $"si-link_{target.PartyUuid:N}_{toPartyUuid:N}_{Guid.NewGuid():N}";
+
+            bool sent = await _notificationClient.SendEmailAsync(
+                target.Email, _settings.EmailSubject, body, idempotencyId, cancellationToken);
```
[the remark this answers is not in the document]
Comment (TheTechArch) [reply src/Authentication/Services/SelfIdentifiedLinkService.cs:79]: Addressed in a3ac3eb — replaced the random GUID with a deterministic minute-bucketed key (`si-link_{from:N}_{to:N}_{yyyyMMddHHmm}`).

I traced this through the Notifications service to confirm it actually works across our (multi-pod) deployment: dedup is enforced by a Postgres `UNIQUE (idempotencyid, creatorname)` constraint on `notifications.orderschain` (migration v0.38). So a deterministic key collapses concurrent/duplicate sends to a single email regardless of which authentication pod handles each click — the constraint is the serialization point, not any per-pod state.

I used a 1-minute window rather than 5 so a genuine resend isn't blocked for long. Real abuse protection (rate-limiting) remains tracked in #2035.

File: src/Authentication/Controllers/SelfIdentifiedAuthenticationController.cs [hunk 2]
```diff
@@ -45,5 +53,64 @@ public async Task<ActionResult> ValidateCredentials([FromBody] SiUserCredentials
...
+            {
+                // The authenticated caller has no party UUID claim, so there is no usable 'to' party to
+                // bind the link to. This is a precondition failure, not a property of the SI username.
+                return BadRequest(new { Message = "Authenticated user has no party UUID." });
+            }
+
+            string? maskedEmail = await _linkService.RequestLinkAsync(request.UserName, toPartyUuid, cancellationToken);
+
+            return Ok(new SelfIdentifiedLinkResponse { MaskedEmail = maskedEmail });
+        }
```
[the remark this answers is not in the document]
Comment (TheTechArch) [reply src/Authentication/Controllers/SelfIdentifiedAuthenticationController.cs:95]: This is a deliberate product decision that came after the original #2035 draft was written. The access-management frontend needs to show the user *where* the link was sent ("Email sent to r*****@g****.com"), so on success the endpoint returns a **masked** address.

This does relax the strict "respond identically" goal. Mitigations in place:
- The endpoint is authenticated (portal scope), so enumeration is limited to logged-in callers, not anonymous.
- Only a **masked** form is ever returned (first char of local part + first char of domain + TLD, fixed-length masks so length isn't revealed) — never the full address.
- The full email never leaves authentication.

Robust abuse protection is rate-limiting (tracked in #2035). I'll update #2035's acceptance criteria so the masked-email return is the documented behavior rather than a conflict.

File: src/Authentication/Controllers/SelfIdentifiedAuthenticationController.cs [hunk 3]
```diff
@@ -45,5 +53,64 @@ public async Task<ActionResult> ValidateCredentials([FromBody] SiUserCredentials
...
+        /// <summary>
+        /// Requests an account-link email for a self-identified user (forgot-password / account-claim
+        /// flow, issue #2035). Looks up the SI user by username and, when it exists with a stored email,
+        /// sends a link (carrying a short-lived token binding the SI user to the authenticated person)
+        /// to that email. Returns the masked recipient address so the caller can show where the email
+        /// went (e.g. "Email sent to r*****@g****.com"); <see cref="SelfIdentifiedLinkResponse.MaskedEmail"/>
+        /// is <c>null</c> when no email was sent.
+        /// </summary>
+        [HttpPost("link-request")]
+        [Authorize(Policy = AuthzConstants.POLICY_SCOPE_PORTAL)]
```
[the remark this answers is not in the document]
Comment (TheTechArch) [reply src/Authentication/Controllers/SelfIdentifiedAuthenticationController.cs:81]: Correct — the scope grew as the flow was built out incrementally over several commits, and the original title/description lagged behind. Updating the PR title and description now to reflect the full delivered scope: link-token mint/validate, SI link-target lookup, link-request (lookup + mint + send via Altinn Notifications), and validate-link-token redemption.

Not splitting: the pieces are tightly coupled (the endpoints depend on the token service and lookup, and share DI/config/tests), so they review and ship as one coherent unit. The redemption-side connection creation in access-management and the remaining infra items (dedicated cert, notifications scope) stay as separate follow-ups noted in #2035.

File: src/Authentication/Configuration/SelfIdentifiedLinkSettings.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    /// Settings for the self-identified account-link request flow (issue #2035): where the emailed
+    /// link points and the email subject.
+    /// </summary>
+    public class SelfIdentifiedLinkSettings
+    {
+        /// <summary>
+        /// Gets or sets the access-management frontend URL that consumes the link token. The token is
+        /// appended as a <c>token</c> query parameter.
+        /// </summary>
+        public string AccessManagementLinkUrl { get; set; } = string.Empty;
```
Comment (Alxandr) [inline src/Authentication/Configuration/SelfIdentifiedLinkSettings.cs]: This should probably be `public Uri?`. Also, using empty string as a default for a value where empty string isn't valid is an anti-pattern.

File: src/Authentication/Controllers/SelfIdentifiedAuthenticationController.cs [hunk 5]
```diff
@@ -38,12 +52,98 @@ public async Task<ActionResult> ValidateCredentials([FromBody] SiUserCredentials
...
+        /// <c>200 Ok(fromPartyUuid)</c> on success, <c>502 Bad Gateway</c> when access-management rejects
+        /// the call. Shared by the credential and link-redemption flows so both delegate identically.
+        /// </summary>
+        private async Task<ActionResult> CreateConnectionAsync(Guid fromPartyUuid, Guid toPartyUuid, CancellationToken cancellationToken)
+        {
+            bool created = await _accessManagementClient.CreateSelfIdentifiedUserConnection(fromPartyUuid, toPartyUuid, cancellationToken);
+
+            if (!created)
+            {
+                return StatusCode(StatusCodes.Status502BadGateway, new { Message = "Failed to create the self-identified user connection." });
```
Comment (Alxandr) [inline src/Authentication/Controllers/SelfIdentifiedAuthenticationController.cs]: Not sure bad-gateway makes sense here? Also, shouldn't this have an error code? This applies to the other errors in this controller as well.

File: src/Authentication/Services/SelfIdentifiedLinkTokenService.cs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+            try
+            {
+                tokenHandler.ValidateToken(token, validationParameters, out SecurityToken validatedToken);
+                JwtSecurityToken jwt = (JwtSecurityToken)validatedToken;
+
+                string? purpose = jwt.Claims.FirstOrDefault(c => c.Type == PurposeClaim)?.Value;
+                if (!string.Equals(purpose, PurposeValue, StringComparison.Ordinal))
+                {
+                    return SelfIdentifiedLinkTokenResult.Invalid("Token purpose mismatch.");
```
Comment (Alxandr) [inline src/Authentication/Services/SelfIdentifiedLinkTokenService.cs]: Instead of taking a string, the `Invalid` function should probably take an enum. Either that, or use `Result` from authorization-utils.

File: src/Authentication/Services/SelfIdentifiedLinkTokenService.cs [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+            }
+        }
+
+        private async Task<X509Certificate2> GetSigningCertificate()
+        {
+            List<X509Certificate2> certificates = await _certificateProvider.GetCertificates();
+
+            X509Certificate2? certificate = certificates
+                .Where(c => c.HasPrivateKey)
+                .OrderByDescending(c => c.NotBefore)
```
Comment (Alxandr) [inline src/Authentication/Services/SelfIdentifiedLinkTokenService.cs:192]: You may want to check that the certificate is actually valid for `now`. So `.Where(c => c.HasPrivateKey && c.NotBefore < now && c.NotAfter > now)` or something like that?

File: src/Core/Models/Profile/SelfIdentifiedLinkRequest.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    /// <summary>
+    /// Request body for the self-identified account-link email endpoint (issue #2035). The
+    /// authenticated person supplies the username of the self-identified account they want to claim.
+    /// </summary>
+    public class SelfIdentifiedLinkRequest
+    {
+        /// <summary>
+        /// Gets or sets the self-identified user's username (login key).
+        /// </summary>
+        public string UserName { get; set; } = string.Empty;
```
Comment (Alxandr) [inline src/Core/Models/Profile/SelfIdentifiedLinkRequest.cs]:
```suggestion
        public string? UserName { get; set; }
```

File: src/Core/Models/Profile/SelfIdentifiedLinkTokenRequest.cs [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    /// <summary>
+    /// Request body for redeeming a self-identified account-link token (issue #2035). Carries the
+    /// token delivered in the account-link email.
+    /// </summary>
+    public class SelfIdentifiedLinkTokenRequest
+    {
+        /// <summary>
+        /// Gets or sets the link token from the email.
+        /// </summary>
+        public string Token { get; set; } = string.Empty;
```
Comment (Alxandr) [inline src/Core/Models/Profile/SelfIdentifiedLinkTokenRequest.cs]:
```suggestion
        public string? Token { get; set; }
```

File: src/Core/Models/Profile/SelfIdentifiedLinkTokenResult.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+        /// <summary>
+        /// Gets the token id (<c>jti</c>), for single-use tracking by the caller.
+        /// </summary>
+        public string? TokenId { get; init; }
+
+        /// <summary>
+        /// Gets a short, non-sensitive reason when <see cref="IsValid"/> is <c>false</c>.
+        /// </summary>
+        public string? Error { get; init; }
```
Comment (Alxandr) [inline src/Core/Models/Profile/SelfIdentifiedLinkTokenResult.cs]: Don't use strings to carry errors. Errors that are returned should have error-codes etc.

File: src/Authentication/Configuration/SelfIdentifiedLinkSettings.cs (hunk 4 above)
Comment (TheTechArch) [reply src/Authentication/Configuration/SelfIdentifiedLinkSettings.cs]: Done in e794617 — changed to `Uri?` with no default. The link service throws a clear `InvalidOperationException` if it's unset rather than silently using an empty string.

File: src/Authentication/Controllers/SelfIdentifiedAuthenticationController.cs (hunk 5 above)
Comment (TheTechArch) [reply src/Authentication/Controllers/SelfIdentifiedAuthenticationController.cs]: Done in e794617 — all failure paths in this controller now return typed problem codes via `ProblemInstance`/`ProblemDescriptor.ToActionResult()` (new `Problem.SelfIdentifiedLink_*` descriptors, AUTH 68–74) instead of anonymous `{ Message }` objects. On the BadGateway question: kept it as a typed `SelfIdentifiedLink_ConnectionFailed` (502) since it represents a failed downstream call to access-management — open to 503/500 if you feel strongly, but it's now a real error code either way.

File: src/Authentication/Services/SelfIdentifiedLinkTokenService.cs (hunk 6 above)
Comment (TheTechArch) [reply src/Authentication/Services/SelfIdentifiedLinkTokenService.cs]: Done in e794617 — dropped the string entirely. `SelfIdentifiedLinkTokenResult.Invalid()` is now parameterless; the validator logs the specific reason internally and the caller just gets a generic invalid-token outcome (mapped to `Problem.SelfIdentifiedLink_InvalidToken`). No reason string is carried on the result.

File: src/Authentication/Services/SelfIdentifiedLinkTokenService.cs (hunk 7 above)
Comment (TheTechArch) [reply src/Authentication/Services/SelfIdentifiedLinkTokenService.cs:192]: Good catch — fixed in e794617. `GetSigningCertificate` now filters `HasPrivateKey && NotBefore.ToUniversalTime() <= now && NotAfter.ToUniversalTime() > now` (now from the injected TimeProvider), so an expired or not-yet-valid cert won't be selected.

File: src/Core/Models/Profile/SelfIdentifiedLinkRequest.cs (hunk 8 above)
Comment (TheTechArch) [reply src/Core/Models/Profile/SelfIdentifiedLinkRequest.cs]: Done in e794617 (`string? UserName`); nullability threaded through the service signatures.

File: src/Core/Models/Profile/SelfIdentifiedLinkTokenRequest.cs (hunk 9 above)
Comment (TheTechArch) [reply src/Core/Models/Profile/SelfIdentifiedLinkTokenRequest.cs]: Done in e794617 (`string? Token`); `ValidateAsync` now takes `string?` and handles null/empty.

File: src/Core/Models/Profile/SelfIdentifiedLinkTokenResult.cs (hunk 10 above)
Comment (TheTechArch) [reply src/Core/Models/Profile/SelfIdentifiedLinkTokenResult.cs]: Done in e794617 — removed the `Error` string from the result entirely (see the token-service thread). Errors are now typed `Problem.SelfIdentifiedLink_*` descriptors returned from the controller.

File: src/Integration/AccessManagement/AccessManagementClient.cs [hunk 11]
```diff
@@ -234,6 +234,36 @@ public async Task<Result<bool>> PushSystemUserToAM(Guid partyUuId, SystemUserInt
...
+                "Authentication // AccessManagementClient // CreateSelfIdentifiedUserConnection // HttpStatusCode: {StatusCode} // {Body}",
+                response.StatusCode,
+                responseContent);
+            return false;
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "Authentication // AccessManagementClient // CreateSelfIdentifiedUserConnection // Exception");
+            throw;
+        }
```
[the remark this answers is not in the document]
Comment (TheTechArch) [reply src/Integration/AccessManagement/AccessManagementClient.cs:278]: Done in e794617 — `CreateSelfIdentifiedUserConnection` now rethrows only on cancellation (`catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)`) and converts other exceptions to `return false`, so transport/runtime failures map to 502 via the normal failure path instead of surfacing a 500.

File: src/Authentication/Controllers/SelfIdentifiedAuthenticationController.cs [hunk 12]
```diff
@@ -13,37 +19,132 @@ namespace Altinn.Platform.Authentication.Controllers
...
-                return Ok(partyUuid);
+                return AuthProblem.SelfIdentifiedLink_InvalidCredentials.ToActionResult();
             }
 
-            return Unauthorized(new { Message = "Invalid credentials." });
+            Guid toPartyUuid = AuthenticationHelper.GetPartyUuId(HttpContext);
+            if (toPartyUuid == Guid.Empty)
+            {
+                return AuthProblem.SelfIdentifiedLink_MissingPartyUuid.ToActionResult();
+            }
```
[the remark this answers is not in the document]
Comment (TheTechArch) [reply src/Authentication/Controllers/SelfIdentifiedAuthenticationController.cs]: Good catch — fixed in 634d41a. Moved the caller party-UUID check to the top of `LinkAccount`, before `ValidateCredentialsAsync`, so a caller without a party UUID is rejected before any failed-attempt/lockout side effect can be recorded on the target SI account. Closes the lockout-abuse vector.

**Review by mgunnerud — APPROVED** [review]

**Review by Alxandr — APPROVED** [review]
